Support SymBool input to torch.compile

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108709
* __->__ #107850
* #107662

We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f, tracing_mode="symbolic")(x)
```

The idea of this PR (credit to @ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305

ydwu4 committed:
Support SymBool input to torch.compiple

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1981,6 +1981,14 @@ def print_results(grouped, indent, result_fn):
+        shape_env.specialize_zero_one = False
+        yield
+    finally:
+        shape_env.specialize_zero_one = prev
```
nuh uh, I'm not sure why you need this but we definitely shouldn't do it this way

ydwu4 replied:
Yeah, I think I find some issue with current PR, will turn into WIP and request for review again.

ydwu4 committed:
Update on "[WIP] Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[WIP] Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[WIP] Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/__init__.py:
```diff
@@ -354,6 +354,14 @@ def __bool__(self):
+        # check equality.
+        if isinstance(other, py_int):  # type: ignore[arg-type]
+            return self.__int__() == other
+        return NotImplemented
```
We should explicitly manage the conversions instead

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -309,6 +309,22 @@ def floordiv(a, b):
+    @staticmethod
+    def expr_cond_pair(a, b):
+        # This handles a pair (expr, cond), where expr is a SymPy expression,
+        # cond represents when it's. The output range is just expr's range.
```
when it's... ?

This seems like a pretty odd thing to have put on value ranges, I would have expected you to have handled this in piecewise all in one go.

ydwu4 replied:
This is kind of implementation detail of piecewise in sympy [here](https://github.com/sympy/sympy/blob/d2be7bacd2604e98a642f74028e8f0d7d6084f78/sympy/functions/elementary/piecewise.py#L137),  which calls this expr_cond_pair inside.  The value ranges for expr_cond_pair will be evaluated before the evaluation of piecewise. So I cound't do it all in one go in piecewise.

Should we keep the current implementation? Or should I figure out a way to avoid adding the expr_cond_pair method.

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -686,6 +686,41 @@ def index_source(key):
                 source=self.source,
                 guards=make_guards(GuardBuilder.FUNCTION_MATCH),
             )
+        elif isinstance(value, torch.SymInt):
```
This seems incomplete? I was expecting you to accept torch.SymBool directly in here, and deal with the int-ify/un-int-ify transparently.

ydwu4 replied:
Yeah, the current implementation is a blindfolded implementation of "need a conversion of the SymBool into a SymInt before it hits the graph".  The graph I understand is dynamo's compiled region. To achieve that, I come up with the idea of  doing the conversion in cond so that both user and dynamo don't need to worry about the conversion.

If we were to detect SymBool then converting it to SymInt inside dynamo, the SymBool inputs will hit the graph. I can explore the implementation a bit.

ezyang commented:
This seems backwards to me. Shouldn't you be detecting SymBool input and converting it to SymInt, rather than what is done here, which is detecting SymInt and (unconditionally!!) converting it to SymBool?

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f)(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f, tracing_mode="symbolic")(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2495,6 +2495,8 @@ def create_symbol(
         constraint_dim: DimConstraint = None,  # NB: includes None
         positive: Optional[bool] = True,
+        *,
+        unspecified: bool = False,
```
Needs some docs distinguishing it from positive. (TBH all of this needs some refactor)

ydwu4 replied:
I can do some refactoring for it.

ezyang commented on test/dynamo/test_export.py:
```diff
@@ -3125,6 +3129,73 @@ def f(x):
+            [true_guard_code, true_guard_code, false_guard_code, false_guard_code],
+            # Outter shape env should have no guards in it because we never specialize on the outter symbool.
+            [[], [], [], []],
+        )
```
So I was expecting to see a test where you make_fx across a torch.cond (that is compiled), and show that you do not specialize the guard in the outer context. Here, you're using torch.export which does a lot more "stuff", and so it's hard for me to tell if the relevant parts actually got exercised.

ydwu4 replied:
it is added in the following PR on the stack https://github.com/pytorch/pytorch/pull/108709 i can move it here to make this PR more self-constained.

ezyang replied:
no point in splitting them, tests with the changes

ezyang replied:
But your tests in that PR don't do any tests on guards / overspec. So it seems like the test is not enough.

ydwu4 replied:
see modified test_cond_with_sym_pred.

ydwu4 commented on torch/_dynamo/source.py:
```diff
@@ -261,6 +261,21 @@ def name(self):
+        return self.base.guard_source()
+
+    def name(self):
+        return f"{self.base.name()}.__int__()"
```
Yeah, one thing is that ITE only accepts and returns boolean. I may need to use piecewise, does it sound good?

ezyang replied:
yeah, whatever lets you get from bool to int symbolically

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f, tracing_mode="symbolic")(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f, tracing_mode="symbolic")(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1240,6 +1240,10 @@ def method_to_operator(method):
         op = getattr(operator, method_attr)
     return op
 
+def convert_to_symint(symbool: torch.SymBool) -> torch.SymInt:
```
Hmm. Under the prevailing naming conventions, we would have called this `sym_int`, but we already have a function named that. Hmm... can we include bool in the name somehow? `cast_symbool_to_symint_guardless` or smthg

ydwu4 replied:
yeah, cast_symbool_to_symint_guardless is accurate.

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f, tracing_mode="symbolic")(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting this, but test_export_with_symbool_inputs is failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/a08e1370ef8cb13cfbf18d9663427a57fa8657f2' -c landrace

ydwu4 committed:
Update on "Support SymBool input to torch.compile"


We could have SymBool inputs for torch.compile, e.g. in the following situation:
```
def f(x:torch.Tensor):
  pred = x.size(0) == 3
  torch.compile(f)(pred, x)

make_fx(f, tracing_mode="symbolic")(x)
```

The idea of this PR (credit to ezyang) is to support SymBool by re-using the infra we've already had for SymInt so that we don't need to replicate a lot of stuff.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng anijain2305

[ghstack-poisoned]
